src/Common/src/CoreWCF/Runtime/TaskHelpers.cs

```diff
@@ -29,120 +30,124 @@ public static async Task AsyncWait<TException>(this Task task)
-                        callbackObj(antecedent);
-                    }, callback, CancellationToken.None, TaskContinuationOptions.HideScheduler, TaskScheduler.Default);
-                }
-                return task;
+                result.ExecuteCallback();
             }
-
-            // Need to create a TaskCompletionSource so that the returned Task object has the correct AsyncState value.
-            var tcs = new TaskCompletionSource<TResult>(state);
-            var continuationState = Tuple.Create(tcs, callback);
-            task.ContinueWith((antecedent, obj) =>
+            else if (callback != null)
             {
-                var tuple = obj as Tuple<TaskCompletionSource<TResult>, AsyncCallback>;
-                var tcsObj = tuple.Item1;
-                var callbackObj = tuple.Item2;
-                if (antecedent.IsFaulted) tcsObj.TrySetException(antecedent.Exception.InnerException);
-                else if (antecedent.IsCanceled) tcsObj.TrySetCanceled();
-                else tcsObj.TrySetResult(antecedent.Result);
-
-                if (callbackObj != null)
+                task.AsTask().ContinueWith((res, asyncResult) =>
                 {
-                    callbackObj(tcsObj.Task);
-                }
-            }, continuationState, CancellationToken.None, TaskContinuationOptions.HideScheduler, TaskScheduler.Default);
-            return tcs.Task;
+                    ((AsyncResult<T>)asyncResult).ExecuteCallback();
+                }, result, CancellationToken.None, TaskContinuationOptions.HideScheduler, TaskScheduler.Default);
+            }
+
+            return result;
         }
 
-        // Helper method when implementing an APM wrapper around a Task based async method which returns a result. 
-        // In the BeginMethod method, you would call use ToApm to wrap a call to MethodAsync:
-        //     return MethodAsync(params).ToApm(callback, state);
-        // In the EndMethod, you would use ToApmEnd to ensure the correct exception handling
-        // This will handle throwing exceptions in the correct place and ensure the IAsyncResult contains the provided
-        // state object
-        public static Task ToApm(this Task task, AsyncCallback callback, object state)
+        /// <summary>
+        /// Helper method to convert from Task async method to "APM" (IAsyncResult with Begin/End calls)
+        /// </summary>
+        public static IAsyncResult ToApm(this Task task, AsyncCallback callback, object state)
         {
-            // When using APM, the returned IAsyncResult must have the passed in state object stored in AsyncState. This
-            // is so the callback can regain state. If the incoming task already holds the state object, there's no need
-            // to create a TaskCompletionSource to ensure the returned (IAsyncResult)Task has the right state object.
-            // This is a performance optimization for this special case.
-            if (task.AsyncState == state)
```

Comment:
The "fast path" is skipped since it never happens where it is used in the framework since state parameters (and the callback) are always used and I don't see why it should change

Reply:
I've seen consumers of APM methods use this pattern:
```c#
  var result = EndFoo(BeginFoo(param, null, null));
```
I know that's a dirty pattern, but I've seen it enough, and actually used it myself in unit tests when I know the underlying implementation is synchronous. Although I'm not sure how important this is now because I believe I've removed every public APM api that I could find in CoreWCF and replaced with Task based ones. If there are any lingering, then they should be removed. So in theory this whole set of methods should be redundant.

Reply:
I did only find the usage at 3 places and as you says it should probably not be used very much, so I don't believe it makes sense to add additional code optimizations which would only apply to tests. 

As for tasks before making the first "stable" release I would recommend you to consider if you maybe should use ValueTask at least for per call methods that are only awaited once. 
I don't have the links at the moment there is a proposal for corefx to make ValueTask 0 allocation (amortized) for simple usage and it is quite good for sync return.

Reply:
I 100% agree. What I first ported this code, ValueTask wasn't a thing yet which is why everything is still a Task. We definitely need to switch to using ValueTask before going stable as that could could be a nasty surprise.